File: algorithms/1_easy/G/00824_goat_latin/solution.cpp
class Solution {
    static inline bool is_vowel(const char& ch) {
        return (
            ch == 'a' || ch == 'e' || ch == 'i' || ch == 'o' || ch == 'u' ||
            ch == 'A' || ch == 'E' || ch == 'I' || ch == 'O' || ch == 'U'
        );
    }
public:
    static string toGoatLatin(const string& sentence) {
        istringstream iss{ sentence };
        ostringstream oss{};
        string word;
        string end{ "maa" };

        while (iss >> word) {
            if (!is_vowel(word[0])) {
                std::rotate(word.begin(), word.begin() + 1, word.end());
            }

            word += end;
            end += "a";

            if (oss.tellp() != 0) {
                oss << ' ';
            }

            oss << word;
        }

        return oss.str();
    }
};

File: algorithms/1_easy/G/00824_goat_latin/solution.rs
impl Solution {

    #[inline(always)]
    fn is_vowel(byte: u8) -> bool {
        matches!(byte, b'a' | b'e' | b'i' | b'o' | b'u' | b'A' | b'E' | b'I' | b'O' | b'U')
    }

    pub fn to_goat_latin(sentence: String) -> String {
        let parts = sentence.split_whitespace().collect::<Vec<_>>();
        let mut end: Vec<u8> = Vec::with_capacity(parts.len() + 3);
        end.extend(b"maa");
        let mut res: Vec<u8> = Vec::with_capacity(
            sentence.len() +
            parts.len() * 2 + 
            (parts.len() * (parts.len() + 1)) / 2
        );

        for part in parts.iter() {
            let mut bytes: Vec<u8> = Vec::with_capacity(part.len() + end.len());
            bytes.extend_from_slice(part.as_bytes());

            if !Self::is_vowel(bytes[0]) {
                bytes.rotate_left(1);
            }

            bytes.extend_from_slice(&end);
            end.push(b'a');

            if !res.is_empty() {
                res.push(b' ')
            }

            res.extend(bytes);
        }

        String::from_utf8(res).unwrap()
    }
}
